piedpiper/gman/client.py

```diff
@@ -0,0 +1,255 @@
+    """
+    Get a list of taskID events from gman.
+    Optionally pass a query_fitler lamda expression to filter these events.
+    For example, to return a list of all failed events for a particular taskID
+    get_task_id_events(
+      task_id='1234', gman_url=url, query_filter=lambda x: x.get('status') == 'failed'
+    )
+    :param task_id: taskID to query as a string
+    :param gman_url: GMan endpoint as as string
+    :param query_filter: lambda expression
+    :return: List of events
+    """
+    try:
+        r = requests.get(f"{gman_url}/task/{task_id}/events")
+        r.raise_for_status()
+    except requests.exceptions.HTTPError as e:
+        raise requests.exceptions.HTTPError(
+            f"Gman returned with a bad status code. \n\n{e}"
+        )
+    except requests.exceptions.RequestException as e:
+        raise requests.exceptions.RequestException(
+            f"Failed to get taskID events for task_id {task_id}. \n\n{e}"
+        )
+    else:
+        if query_filter:
+            return list(filter(query_filter, r.json()))
+        else:
+            return r.json()
+
+
+def get_thread_id_tasks(thread_id=None, gman_url=None, query_filter=None):
+    """
+    Get a list of tasks associated with the given thread_id
+    :param thread_id: The thread_id to query with as a sring
+    :param gman_url: The GMan endpoint as a string
+    :param query_filter: lambda expression
+    :return: List of tasks associated with thread_id
+    """
+    try:
+        r = requests.get(f"{gman_url}/thread/{thread_id}")
+        r.raise_for_status()
+    except requests.exceptions.HTTPError as e:
+        raise requests.exceptions.HTTPError(
+            f"Gman returned with a bad status code. \n\n{e}"
+        )
+    except requests.exceptions.RequestException as e:
+        raise requests.exceptions.RequestException(
+            f"Failed to get tasks for thread_id {thread_id}. \n\n{e}"
+        )
+    else:
+        if query_filter:
+            return list(filter(query_filter, r.json()))
+        else:
+            return r.json()
+
+
+def get_thread_id_events(thread_id=None, gman_url=None, query_filter=None):
+    """
+    Get list of all events for a given thread_id, optionally filtered by
+    a lambda expression.
+    :param thread_id: Thread ID to query for as a string.
+    :param gman_url:  GMan endpoint as a string.
+    :param query_filter: Lambda expression
+    :return: List of task events
+    """
+    try:
+        r = requests.get(f"{gman_url}/thread/{thread_id}/events")
+        r.raise_for_status()
+    except requests.exceptions.HTTPError as e:
+        raise requests.exceptions.HTTPError(
+            f"Gman returned with a bad status code. \n\n{e}"
+        )
+    except requests.exceptions.RequestException as e:
+        raise requests.exceptions.RequestException(
+            f"Failed to get tasks for thread_id {thread_id}. \n\n{e}"
+        )
+    else:
```

Comment:
I think this should be in try. else is usually reserved for handling things that aren't supposed to generate an exception. So in this case if an exception is thrown in your r handling code... it would propagate up.

Additionally try: except: else: logic is not well understood by most devs and therefor might be more prone to unexpected/unintended behaviors as well as a more complex logic flow.

Most of the time logic handling data set that might throw an exception handled as if there were no try: except: block one after another, separating seems less logically direct and could increase the likelihood of missing logic that is connected to "r" in this case. E.G. changing r to another variable could be missed in the else: or other more obscure changes to the try: block might be missed in an else: